### PR TITLE
Enhance MemorySchemaProvider to support `register_listing_table`

### LINKED ISSUE #1836: Register multiple tables into `ExecutionContext` at once
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

I want to make it really easy to register multiple tables in a directory into a `ExecutionContext` so I can write SQL queries on them.

**Describe the solution you'd like**

For example if their is a data directory like the below:

s3://bucket/active/table1/abc.parquet
s3://bucket/active/table2/abc.parquet
s3://bucket/active/table3/abc.parquet
s3://bucket/historical/table1/abc.parquet
s3://bucket/historical/table2/abc.parquet
s3://bucket/historical/table3/abc.parquet

I would like to be able to do something like the below.

```
ctx.register_object_store("s3", object_store);

// Provide a list of tables and table name prefix
// Registers active_table1, active_table2 into ctx
let tables = vec!["table1", "table2"];
ctx.register_listing_tables(
    object_store: "s3", 
    object_store_prefix: "bucket/active", 
    tables: Some(&tables),
    table_prefix: Some("active")
);

// Provide a prefix and register everything in the level below that as a table with a table name prefix
// Registers active_table1, active_table2, active_table3 into ctx
ctx.register_listing_tables(
    object_store: "s3", 
    object_store_prefix: "bucket/active", 
    tables: None
    table_prefix: "active"
);

// Provide a prefix and register everything in the level below that as a table with a no prefix
// Registers table1, table2, table3 into ctx
ctx.register_listing_tables(
    object_store: "s3", 
    object_store_prefix: "bucket/active", 
    tables: None
    table_prefix: None
);
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Closes #1836 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
Im looking into whether I should just have `register_object_store` do the same thing as the same named method on `ExecutionContext`

**Comment by matthewmturner** [comment]
@alamb @seddonm1 @Igosuki @returnString i've extended the `MemorySchemaProvider` to support creating `ListingTable` from an `ObjectStore`.  This can further be used with `object_store.list_file(...)` to create many tables in a directory at once.

I havent yet tested with partitioned data, but I think this is a good start.

I'm interested in your views.

**Comment by matthewmturner** [comment]
@houqp do you have any thoughts on this?

**Comment by matthewmturner** [comment]
Its also quite interesting how this only fails CI tests on windows.

Has anyone seen that issue with the object store abstraction before?

**Comment by alamb** [comment]
Sorry for the late review. Looking now

**Comment by matthewmturner** [comment]
@alamb do you think `ObjectStoreSchemaProvider` is the way to go here? Doing that makes sense to me to keep a cleaner separation of functionality without the bloat of `ObjectStore` when it isn't needed.

**Comment by alamb** [comment]
> @alamb do you think ObjectStoreSchemaProvider is the way to go here? Doing that makes sense to me to keep a cleaner separation of functionality without the bloat of ObjectStore when it isn't needed.

Yes, I do think that would be a cleaner separation of functionality. Does anyone else have opinions (@rdettai  or @yjshen  for example?)

**Comment by Igosuki** [comment]
I don't understand why there should be various interfaces. If you want to achieve abstraction over tables, the best is to enable a single data abstraction/discovery interface that points to tables no matter where they are ?

**Comment by alamb** [comment]
> I don't understand why there should be various interfaces. If you want to achieve abstraction over tables, the best is to enable a single data abstraction/discovery interface that points to tables no matter where they are ?


I think it comes down to if the InMemoryCatalog implementation would be used for interfaces on top of files -- aka do we want all InMemoryCatalog's to know about ObjectStores? 

I can go either way -- if people prefer this way I am fine with that -- let's get the PR polished up and we can merge it in

**Comment by matthewmturner** [comment]
I agree with @Igosuki that it would be nice to have a single abstraction but im not sure this is the PR for that.  I've seen talks of other table providers, for example on slack a BigTable table provider was mentioned / potentially in the works - so its not clear to me where we would want to draw the line yet.

My proposal would be to get this merged with the clear separation of functionalities (i.e. i migrate to `ObjectStoreSchemaProvider` and I can also make a separate issue to discuss having a single interface - for example maybe a `TableRegistry`.

**Comment by alamb** [comment]
> My proposal would be to get this merged with the clear separation of functionalities (i.e. i migrate to ObjectStoreSchemaProvider and I can also make a separate issue to discuss having a single interface.


Sounds good to me -- it seems to me like the only blocker is whatever is causing the test failure on Windows

File: datafusion/src/catalog/schema.rs [hunk 1]
```diff
@@ -154,4 +208,100 @@ mod tests {
...
+        let store = schema
+            .register_object_store("file", Arc::new(LocalFileSystem {}))
+            .unwrap();
+
+        let mut files = store.list_file(&testdata).await.unwrap();
+        while let Some(file) = files.next().await {
+            let sized_file = file.unwrap().sized_file;
+            let file = sized_file.path.split('/').last().unwrap();
+            if file == "alltypes_dictionary.parquet" || file == "alltypes_plain.parquet" {
+                let (name, _) = file.split_once(".").unwrap();
```
Comment (yjshen) [inline datafusion/src/catalog/schema.rs:270]:
```suggestion
            let path = Path::new(&sized_file.path);
            let file = path.file_name().unwrap();
            if file == OsStr::new("alltypes_dictionary.parquet")
                || file == OsStr::new("alltypes_plain.parquet")
            {
                let name = path.file_stem().unwrap().to_str().unwrap();
```

Comment (matthewmturner) [reply datafusion/src/catalog/schema.rs]: @yjshen thx much

**Comment by matthewmturner** [comment]
@alamb hopefully this is good now.

**Review by alamb — APPROVED** [review]
Thanks @matthewmturner and @jychen7

**Pull request merged**